src/cmap/auth/mongodb_oidc.ts

```diff
@@ -1,39 +1,136 @@
   tokenEndpoint?: string;
-  deviceAuthorizeEndpoint?: string;
+  deviceAuthorizationEndpoint?: string;
   clientId: string;
   clientSecret?: string;
   requestScopes?: string[];
 }
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export interface OIDCRequestTokenResult {
   accessToken: string;
   expiresInSeconds?: number;
   refreshToken?: string;
 }
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export type OIDCRequestFunction = (
-  idl: OIDCMechanismServerStep1
+  principalName: string,
+  serverResult: OIDCMechanismServerStep1,
+  timeout: AbortSignal | number
 ) => Promise<OIDCRequestTokenResult>;
 
-/**
- * TODO: NODE-5035: Make API public
- *
- * @internal */
+/** @public */
 export type OIDCRefreshFunction = (
-  idl: OIDCMechanismServerStep1,
-  result: OIDCRequestTokenResult
+  principalName: string,
+  serverResult: OIDCMechanismServerStep1,
+  result: OIDCRequestTokenResult,
+  timeout: AbortSignal | number
 ) => Promise<OIDCRequestTokenResult>;
+
+/** @internal */
+export const OIDC_WORKFLOWS = {
+  callback: new CallbackWorkflow(),
+  aws: new AwsDeviceWorkflow(),
+  azure: undefined,
+  gcp: undefined,
+  // Set the protype to null to avoid false returns and keep types.
+  __proto__: null
+};
+
+/**
+ * OIDC auth provider.
+ */
+export class MongoDBOIDC extends AuthProvider {
+  /**
+   * Instantiate the auth provider.
+   */
+  constructor() {
+    super();
+  }
+
+  /**
+   * Authenticate using OIDC
+   */
+  override auth(authContext: AuthContext, callback: Callback): void {
+    const { connection, credentials, response } = authContext;
+
+    if (response && response.speculativeAuthenticate) {
+      return callback();
+    }
+
+    if (!credentials) {
+      return callback(new MongoMissingCredentialsError('AuthContext must provide credentials.'));
+    }
+
+    getWorkflow(credentials, (error, workflow) => {
+      if (error || !workflow) {
+        return callback(error);
+      }
+      workflow
+        .execute(connection, credentials)
+        .then(result => {
+          return callback(undefined, result);
+        })
+        .catch(error => {
+          callback(error);
+        });
+    });
+  }
+
+  /**
+   * Add the speculative auth for the initial handshake.
+   */
+  override prepare(
+    handshakeDoc: HandshakeDocument,
+    authContext: AuthContext,
+    callback: Callback<HandshakeDocument>
+  ): void {
+    const { credentials } = authContext;
+
+    if (!credentials) {
+      return callback(new MongoMissingCredentialsError('AuthContext must provide credentials.'));
+    }
+
+    getWorkflow(credentials, (error, workflow) => {
+      if (error || !workflow) {
+        return callback(error);
+      }
+      workflow
+        .speculativeAuth()
+        .then(result => {
+          return callback(undefined, { ...handshakeDoc, ...result });
+        })
+        .catch(error => {
+          callback(error);
+        });
```

Comment:
This applies to all `then` chaining.

We should use the overload `.then(<success cb>, <error cb>)`.  If we use a `.then().catch()`, if `callback` throws in the `then` block, the callback will be called a second time in the catch block when that error is caught.

We pass this callback in so we can guarantee it never throws, but as a best practice we should opt for the overload that takes both the then and catch handlers in a single overload.

Reply:
Updated - thanks for catching that.